File: 2term/jvmWorld/1lab_scala_binaryTree/src/main/scala/GeoCoordinate.scala
import scala.util.Try

class GeoCoordinate(val latitude: Double, val longitude: Double) {
  override def toString: String = {
    s"Lat: $latitude, Long: $longitude"
    //    s"Latitude: $latitude, Longitude: $longitude"
  }

  def serializeToString: String = {
    s"$latitude|$longitude"
  }


}

object GeoCoordinate {
  val strname = "GeoCoordinate"
  def deserializeFromString(str: String): Option[GeoCoordinate] = {
    str.split('|') match {
      case Array(latStr, lonStr) => Try {
        val latitude = latStr.toDouble
        val longitude = lonStr.toDouble
        Some(new GeoCoordinate(latitude, longitude))
      }.getOrElse(None)
      case _ => None
    }
  }

//  def strname: String = "GeoCoordinate"
}

class GeoCoordinateLatitudeComparator extends Ordering[GeoCoordinate] {
  override def compare(a: GeoCoordinate, b: GeoCoordinate): Int = {
    if (a.latitude == b.latitude) {
      0
    }
    else if (a.latitude > b.latitude) {
      1
    }
    else {
      -1
    }
  }

  override def equiv(x: GeoCoordinate, y: GeoCoordinate): Boolean = {
    compare(x, y) == 0
  }

  override def lt(x: GeoCoordinate, y: GeoCoordinate): Boolean = {
    compare(x, y) == -1
  }

  override def gt(x: GeoCoordinate, y: GeoCoordinate): Boolean = {
    compare(x, y) == 1
  }
}

class GeoCoordinateLongitudeComparator extends Ordering[GeoCoordinate] {
  override def compare(a: GeoCoordinate, b: GeoCoordinate): Int = {
    if (a.longitude == b.longitude) {
      0
    }
    else if (a.longitude > b.longitude) {
      1
    }
    else {
      -1
    }
  }

  override def equiv(x: GeoCoordinate, y: GeoCoordinate): Boolean = {
    compare(x, y) == 0
  }

  override def lt(x: GeoCoordinate, y: GeoCoordinate): Boolean = {
    compare(x, y) == -1
  }

  override def gt(x: GeoCoordinate, y: GeoCoordinate): Boolean = {
    compare(x, y) == 1
  }
}

File: 2term/jvmWorld/MyApplication/app/src/main/java/com/example/myapplication/GeoCoordinate.kt
package com.example.myapplication

import kotlin.random.Random
import java.math.RoundingMode

class GeoCoordinate(val latitude: Double, val longitude: Double) : CustomTypeInterface {

    override val typeName: String
        get() = "GeoCoordinate"

    override val typeExampleRepresentation: String
        get() = "1.5|2.8"

    override fun createRandomInstance(): CustomTypeInterface {
        val latitude = (Random.nextDouble() * 200 - 100).toBigDecimal().setScale(2, RoundingMode.HALF_UP).toDouble()
        val longitude = (Random.nextDouble() * 200 - 100).toBigDecimal().setScale(2, RoundingMode.HALF_UP).toDouble()

        return GeoCoordinate(latitude, longitude)
    }

    override fun toString(): String {
        return "Lat: $latitude, Long: $longitude"
    }

    override fun serializeToString(): String {
        return "$latitude|$longitude"
    }

    override fun deserializeFromString(str: String): CustomTypeInterface? {
        val parts = str.split('|')
        return when {
            parts.size == 2 -> {
                try {
                    val latitude = parts[0].toDouble()
                    val longitude = parts[1].toDouble()
                    GeoCoordinate(latitude, longitude)
                } catch (e: NumberFormatException) {
                    null
                }
            }
            else -> null
        }
    }

    override fun compare(other: CustomTypeInterface): Int {
        return when (other) {
            is GeoCoordinate -> {
                when {
                    latitude < other.latitude -> -1
                    latitude > other.latitude -> 1
                    else -> 0
                }
            }
            else -> throw IllegalArgumentException("Cannot compare GeoCoordinate with a different type")
        }
    }
}
